// colorizr/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.textColor = exports.scheme = exports.saturate = exports.rotate = exports.rgb2hsl = exports.rgb2hex = exports.random = exports.parseCSS = exports.palette = exports.name = exports.luminance = exports.lighten = exports.isValidHex = exports.isValidColor = exports.hsl2rgb = exports.hsl2hex = exports.hex2rgb = exports.hex2hsl = exports.formatHex = exports.formatCSS = exports.fade = exports.desaturate = exports.darken = exports.contrast = exports.compare = exports.colorDifference = exports.chroma = exports.brightnessDifference = void 0;
const chroma_1 = require("./chroma");
const compare_1 = require("./compare");
const darken_1 = require("./darken");
const desaturate_1 = require("./desaturate");
const fade_1 = require("./fade");
const format_css_1 = require("./format-css");
const lighten_1 = require("./lighten");
const luminance_1 = require("./luminance");
const parse_color_1 = require("./modules/parse-color");
const utils_1 = require("./modules/utils");
const rotate_1 = require("./rotate");
const saturate_1 = require("./saturate");
const text_color_1 = require("./text-color");
class Colorizr {
    hex;
    model;
    hsl;
    rgb;
    constructor(color, options = {}) {
        (0, utils_1.invariant)(!!color, 'color is required');
        const { model = 'rgb' } = options;
        const { hex, hsl, rgb } = (0, parse_color_1.default)(color);
        this.model = model;
        this.hex = hex;
        this.hsl = hsl;
        this.rgb = rgb;
    }
    /**
     * Get css string
     */
    get css() {
        return (0, format_css_1.default)(this.hsl, { model: this.model });
    }
    /**
     * Get the red value
     */
    get red() {
        return Number(this.rgb.r);
    }
    /**
     * Get the green value
     */
    get green() {
        return Number(this.rgb.g);
    }
    /**
     * Get the blue value
     */
    get blue() {
        return Number(this.rgb.b);
    }
    /**
     * Get the hue value
     */
    get hue() {
        return Number(this.hsl.h);
    }
    /**
     * Get the saturation value
     */
    get saturation() {
        return Number(this.hsl.s);
    }
    /**
     * Get the lightness value
     */
    get lightness() {
        return Number(this.hsl.l);
    }
    /**
     * Get the luminance value
     */
    get luminance() {
        return (0, luminance_1.default)(this.hex);
    }
    /**
     * Get the chroma value
     */
    get chroma() {
        return (0, chroma_1.default)(this.hex);
    }
    /**
     * Get the contrasted color
     */
    get textColor() {
        return (0, text_color_1.default)(this.hex);
    }
    /**
     * Test 2 colors for compliance
     */
    compare(input) {
        return (0, compare_1.default)(this.hex, input);
    }
    /**
     * Increase lightness
     */
    lighten(percentage = 10) {
        return (0, lighten_1.default)(this.hex, percentage);
    }
    /**
     * Decrease lightness
     */
    darken(percentage = 10) {
        return (0, darken_1.default)(this.hex, percentage);
    }
    /**
     * Increase saturation
     */
    saturate(percentage = 10) {
        return (0, saturate_1.default)(this.hex, percentage);
    }
    /**
     * Decrease saturation
     */
    desaturate(percentage = 10) {
        return (0, desaturate_1.default)(this.hex, percentage);
    }
    /**
     * Invert color
     */
    invert() {
        return (0, rotate_1.default)(this.hex, 180);
    }
    /**
     * Rotate color
     */
    rotate(degrees = 15) {
        return (0, rotate_1.default)(this.hex, degrees);
    }
    /**
     * Fade color
     */
    fade(percentage = 10) {
        return (0, fade_1.default)(this.hex, percentage, this.model);
    }
}
var brightness_difference_1 = require("./brightness-difference");
Object.defineProperty(exports, "brightnessDifference", { enumerable: true, get: function () { return brightness_difference_1.default; } });
var chroma_2 = require("./chroma");
Object.defineProperty(exports, "chroma", { enumerable: true, get: function () { return chroma_2.default; } });
var color_difference_1 = require("./color-difference");
Object.defineProperty(exports, "colorDifference", { enumerable: true, get: function () { return color_difference_1.default; } });
var compare_2 = require("./compare");
Object.defineProperty(exports, "compare", { enumerable: true, get: function () { return compare_2.default; } });
var contrast_1 = require("./contrast");
Object.defineProperty(exports, "contrast", { enumerable: true, get: function () { return contrast_1.default; } });
var darken_2 = require("./darken");
Object.defineProperty(exports, "darken", { enumerable: true, get: function () { return darken_2.default; } });
var desaturate_2 = require("./desaturate");
Object.defineProperty(exports, "desaturate", { enumerable: true, get: function () { return desaturate_2.default; } });
var fade_2 = require("./fade");
Object.defineProperty(exports, "fade", { enumerable: true, get: function () { return fade_2.default; } });
var format_css_2 = require("./format-css");
Object.defineProperty(exports, "formatCSS", { enumerable: true, get: function () { return format_css_2.default; } });
var format_hex_1 = require("./format-hex");
Object.defineProperty(exports, "formatHex", { enumerable: true, get: function () { return format_hex_1.default; } });
var hex2hsl_1 = require("./hex2hsl");
Object.defineProperty(exports, "hex2hsl", { enumerable: true, get: function () { return hex2hsl_1.default; } });
var hex2rgb_1 = require("./hex2rgb");
Object.defineProperty(exports, "hex2rgb", { enumerable: true, get: function () { return hex2rgb_1.default; } });
var hsl2hex_1 = require("./hsl2hex");
Object.defineProperty(exports, "hsl2hex", { enumerable: true, get: function () { return hsl2hex_1.default; } });
var hsl2rgb_1 = require("./hsl2rgb");
Object.defineProperty(exports, "hsl2rgb", { enumerable: true, get: function () { return hsl2rgb_1.default; } });
var is_valid_color_1 = require("./is-valid-color");
Object.defineProperty(exports, "isValidColor", { enumerable: true, get: function () { return is_valid_color_1.default; } });
var is_valid_hex_1 = require("./is-valid-hex");
Object.defineProperty(exports, "isValidHex", { enumerable: true, get: function () { return is_valid_hex_1.default; } });
var lighten_2 = require("./lighten");
Object.defineProperty(exports, "lighten", { enumerable: true, get: function () { return lighten_2.default; } });
var luminance_2 = require("./luminance");
Object.defineProperty(exports, "luminance", { enumerable: true, get: function () { return luminance_2.default; } });
var name_1 = require("./name");
Object.defineProperty(exports, "name", { enumerable: true, get: function () { return name_1.default; } });
var palette_1 = require("./palette");
Object.defineProperty(exports, "palette", { enumerable: true, get: function () { return palette_1.default; } });
var parse_css_1 = require("./parse-css");
Object.defineProperty(exports, "parseCSS", { enumerable: true, get: function () { return parse_css_1.default; } });
var random_1 = require("./random");
Object.defineProperty(exports, "random", { enumerable: true, get: function () { return random_1.default; } });
var rgb2hex_1 = require("./rgb2hex");
Object.defineProperty(exports, "rgb2hex", { enumerable: true, get: function () { return rgb2hex_1.default; } });
var rgb2hsl_1 = require("./rgb2hsl");
Object.defineProperty(exports, "rgb2hsl", { enumerable: true, get: function () { return rgb2hsl_1.default; } });
var rotate_2 = require("./rotate");
Object.defineProperty(exports, "rotate", { enumerable: true, get: function () { return rotate_2.default; } });
var saturate_2 = require("./saturate");
Object.defineProperty(exports, "saturate", { enumerable: true, get: function () { return saturate_2.default; } });
var scheme_1 = require("./scheme");
Object.defineProperty(exports, "scheme", { enumerable: true, get: function () { return scheme_1.default; } });
var text_color_2 = require("./text-color");
Object.defineProperty(exports, "textColor", { enumerable: true, get: function () { return text_color_2.default; } });
__exportStar(require("./types"), exports);
exports.default = Colorizr;
